geode-core/src/main/java/org/apache/geode/distributed/internal/FunctionExecutionPooledExecutor.java

```diff
@@ -86,15 +96,32 @@ public void rejectedExecution(final Runnable r, ThreadPoolExecutor executor) {
             throw new RejectedExecutionException(
                 "executor has been shutdown");
           } else {
-            // System.out.println("Asif: Rejection called");
-            if (Thread
-                .currentThread() == ((FunctionExecutionPooledExecutor) executor).bufferConsumer) {
-              Thread th = executor.getThreadFactory().newThread((new Runnable() {
-                public void run() {
-                  r.run();
+            // The request was rejected due to all threads being in use.
+            //
+            // In the normal case, add the rejected request to the blocking queue.
+            //
+            // If the handler was invoked via the bufferConsumer (the thread that takes from the
+            // blocking queue and offers to the synchronous queue), then spin off a thread directly.
+            // In this case, an offer has already been made to the synchronous queue and failed, so
+            // all the function execution threads are in use.
+            //
+            // If the handler was invoked via a function execution thread, then spin off a thread
+            // directly. In this case, that means a function is executing another function. The
+            // child function request shouldn't be in the queue behind the parent request since the
+            // parent function is dependent on the child function executing.
+            boolean isBufferConsumer = isBufferConsumer(executor);
```

Comment:
Could be made `final`

Reply:
Honestly, I think using final on local variables really just adds a lot of noise. People used to claim that it sped up the code, but I've never seen that proven, plus the JIT figures things out so fast that any improvement would be very short-lived.